File: rdf-infographic-skill/scripts/generate_infographic.py
#!/usr/bin/env python3
"""
RDF Infographic Generator — Build self-contained HTML infographics from RDF data.

Usage:
    python3 generate_infographic.py --ttl input.ttl --output out.html
    python3 generate_infographic.py --ttl input.ttl --output out.html --title "My Title" --tagline "..." --source-url "..."

The generator parses the RDF, extracts entities and narrative sections,
then assembles a complete HTML infographic with:
  - Floating navigation with theme toggle
  - D3.js Knowledge Graph Explorer (Basic/Advanced modes)
  - FAQ, glossary, HowTo sections (auto-extracted from RDF annotations)
  - SPARQL query workbench
  - Attribution footer
"""

from __future__ import annotations
import argparse
import sys
from pathlib import Path

HERE = Path(__file__).parent
sys.path.insert(0, str(HERE))

from html_assembler import assemble_html


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Generate RDF-backed HTML infographic",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )
    parser.add_argument("--ttl", required=True, help="Input RDF file (Turtle/JSON-LD/NTriples)")
    parser.add_argument("--output", "-o", required=True, help="Output HTML file path")
    parser.add_argument("--title", default="", help="Page title (default: derived from RDF)")
    parser.add_argument("--description", default="", help="Meta description")
    parser.add_argument("--tagline", default="", help="Hero tagline text")
    parser.add_argument("--hero-tagline", default="", help="Hero metadata line (e.g. author + source)")
    parser.add_argument("--meta-html", default="", help="Hero meta HTML (published date, source link)")
    parser.add_argument("--source-url", default="", help="Source URL for attribution")
    parser.add_argument("--source-label", default="", help="Source label for attribution")
    parser.add_argument("--resolver-pattern", default="https://linkeddata.uriburner.com/describe/?url=",
                        help="Resolver URL pattern (default: URIBurner)")

    args = parser.parse_args()

    rdf_path = Path(args.ttl)
    if not rdf_path.exists():
        print(f"Error: RDF file not found: {rdf_path}", file=sys.stderr)
        return 1

    success = assemble_html(
        rdf_path=args.ttl,
        output_path=args.output,
        title=args.title,
        description=args.description,
        tagline=args.tagline,
        hero_tagline=args.hero_tagline,
        meta_html=args.meta_html,
        source_url=args.source_url,
        source_label=args.source_label,
        resolver_pattern=args.resolver_pattern,
    )

    return 0 if success else 1


if __name__ == "__main__":
    sys.exit(main())

File: rdf-infographic-skill/scripts/generate_infographic.ts
/**
 * RDF Infographic Generator — TypeScript edition (Node.js ≥ 18, requires n3).
 * Builds self-contained HTML infographics from RDF data.
 * Mirrors generate_infographic.py — same CLI flags, same behavior.
 *
 * Usage:
 *   npx tsx generate_infographic.ts --ttl input.ttl --output out.html
 *   npx tsx generate_infographic.ts --ttl input.ttl -o out.html --title "My Title" --tagline "..."
 */

import { existsSync } from "node:fs";
import { assembleHtml, type AssembleHtmlOptions } from "./html_assembler.ts";

interface Args {
  ttl?: string;
  output?: string;
  title: string;
  description: string;
  tagline: string;
  heroTagline: string;
  metaHtml: string;
  sourceUrl: string;
  sourceLabel: string;
  resolverPattern: string;
}

function parseArgs(argv: string[]): Args {
  const args: Args = {
    title: "",
    description: "",
    tagline: "",
    heroTagline: "",
    metaHtml: "",
    sourceUrl: "",
    sourceLabel: "",
    resolverPattern: "https://linkeddata.uriburner.com/describe/?url=",
  };
  for (let i = 0; i < argv.length; i++) {
    switch (argv[i]) {
      case "--ttl":               args.ttl          = argv[++i]; break;
      case "--output": case "-o": args.output       = argv[++i]; break;
      case "--title":             args.title         = argv[++i]; break;
      case "--description":       args.description   = argv[++i]; break;
      case "--tagline":           args.tagline        = argv[++i]; break;
      case "--hero-tagline":      args.heroTagline    = argv[++i]; break;
      case "--meta-html":         args.metaHtml       = argv[++i]; break;
      case "--source-url":        args.sourceUrl      = argv[++i]; break;
      case "--source-label":      args.sourceLabel    = argv[++i]; break;
      case "--resolver-pattern":  args.resolverPattern = argv[++i]; break;
    }
  }
  return args;
}

function main(): number {
  const args = parseArgs(process.argv.slice(2));

  if (!args.ttl) {
    process.stderr.write("Error: --ttl is required\n");
    process.stderr.write(
      "Usage: npx tsx generate_infographic.ts --ttl input.ttl --output out.html\n",
    );
    return 1;
  }
  if (!args.output) {
    process.stderr.write("Error: --output/-o is required\n");
    return 1;
  }
  if (!existsSync(args.ttl)) {
    process.stderr.write(`Error: RDF file not found: ${args.ttl}\n`);
    return 1;
  }

  const opts: AssembleHtmlOptions = {
    rdfPath:         args.ttl,
    outputPath:      args.output,
    title:           args.title,
    description:     args.description,
    tagline:         args.tagline,
    heroTagline:     args.heroTagline,
    metaHtml:        args.metaHtml,
    sourceUrl:       args.sourceUrl,
    sourceLabel:     args.sourceLabel,
    resolverPattern: args.resolverPattern,
  };

  const success = assembleHtml(opts);
  return success ? 0 : 1;
}

if (
  process.argv[1] &&
  (process.argv[1].endsWith("generate_infographic.ts") ||
    process.argv[1].endsWith("generate_infographic.js"))
) {
  process.exit(main());
}
